clock_control: stm32: Configure MCO through devicetree instead of Kconfig

joerchan opened the pull request:
Deprecated setting MCO source and prescalor in Kconfig in favour of setting it through devicetree.

joerchan committed:
drivers: Add driver for STM32 MCO peripheral

Add device driver for STM32 MCO peripheral which takes configures
the MCO clock source and prescaler, and outputs it on one of the GPIO
pins.

Signed-off-by: Joakim Andersson <joerchan@gmail.com>

joerchan committed:
dts: st: Add MCO node to STM32 boards

Add MCO device nodes to the STM32 boards.
The set of supported boards are chosen to replace what is currently
supported in Kconfig.

Signed-off-by: Joakim Andersson <joerchan@gmail.com>

joerchan committed:
samples: stm32: mco: Update sample to use dts instead of Kconfig

Update MCO board sample to use devicetree information instead of Kconfig
configurations.

Signed-off-by: Joakim Andersson <joerchan@gmail.com>

joerchan committed:
drivers: clock_control: Deprecate Kconfig for MCO configuration

Deprecate support for configuring the MCO source and prescaler from
Kconfig configurations.
This is now done by devicetree and an MCO driver instead, which also
configures the pin to be used by the MCO peripheral.

Signed-off-by: Joakim Andersson <joerchan@gmail.com>

joerchan committed:
dt-bindings: clock: Add clock sources for stm32f1x/10x for MCO

Add clock sources that can be output by the MCO on the stm32f1x and
stm32f10 connectivity line devices.

Signed-off-by: Joakim Andersson <joerchan@gmail.com>

joerchan commented on samples/boards/stm32/mco/boards/nucleo_u5a5zj_q.overlay:
```diff
@@ -0,0 +1,17 @@
+	status = "okay";
+};
+
+&clk_hse {
```
Will remove this again, was used for testing.

mathieuchopstm commented on drivers/clock_control/clock_stm32_mco.c:
```diff
@@ -0,0 +1,136 @@
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(mco1), okay)
+#if DT_PROP(DT_NODELABEL(mco1), source) == NO_SEL
+	#define MCO1_SOURCE		LL_RCC_MCO1SOURCE_NOCLOCK
+#elif DT_PROP(DT_NODELABEL(mco1), source) == STM32_SRC_EXT_HSE
```
Since sources may not be defined, I think all these checks should be of the form:
```suggestion
#elif defined(STM32_SRC_EXT_HSE) && DT_PROP(DT_NODELABEL(mco1), source) == STM32_SRC_EXT_HSE
```

joerchan replied:
This part of the code was removed.

mathieuchopstm commented on dts/bindings/clock/st,stm32-clock-mco.yaml:
```diff
@@ -0,0 +1,43 @@
+        Configure the microcontroller clock output pin and additional properties,
+        e.g. clock source and prescaler.
+        Example:
+                &mco {
```
Nit: since the node is always named `mco1`
```suggestion
                &mco1 {
```

mathieuchopstm commented on include/zephyr/dt-bindings/clock/stm32f1_clock.h:
I feel like this may be out of place - however, I see the benefit of adding it to per-series DT bindings header (only the supported MCO sources are listed).

I would be fine with this if the defines were renamed and placed in a "separate" list, e.g.:
```c
/** MCO source clocks */

#define STM32_MCO_SRC_HSE 0
#define STM32_MCO_SRC_LSE (STM32_MCO_SRC_HSE + 1)
        /* etc... */
```
we could even place the common definitions in `stm32_common_clocks.h` as done for domain clocks.

@erwango what do you think?

erwango replied:
I don't see the immediate benefit of splitting in a new list, tbh.
On the contrary, reusing existing sources definitions would ensure some minimal consistency, IMO.

mathieuchopstm replied:
I was thinking about splitting because the `STM32_SRC_xxx` sources are accepted as `pclken.bus` for calls to `clock_control_get_subsys_rate` queries, whereas these would not be (although support for them could be added).

I guess it's not too big of a risk though.

However, if we keep in the list as-is, shouldn't the `SRC_EXT_HSE` (for example) be dropped, and the regular `SRC_HSE` being used instead?

joerchan replied:
Given the current change-set I don't think there is anything for my to do about this comment now.

mathieuchopstm replied:
Yes, marking conversation as resolved.

mathieuchopstm commented on samples/boards/stm32/mco/README.rst:
```diff
@@ -0,0 +1,29 @@
+Overview
+********
+
+This sample is a minimum application to demonstrate how to output one of the internal clocks for
+external use by the application.
```
Nit: align with other Zephyr samples (+ rewording)
```suggestion
This sample demonstrates how to use the Microcontroller Clock Output (MCO) capability of STM32 microcontrollers.

MCO can be used to output an internally generated clock signal on a GPIO pin of the MCU, for usage by external peripherals.
```

joerchan replied:
The sample is already merged. Not part of this review.

mathieuchopstm commented on drivers/clock_control/clock_stm32_mco.c:
```diff
@@ -0,0 +1,136 @@
+	.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(inst),                   \
+};                                                                      \
+static struct stm32_mco_data stm32_mco_data_##inst = {                  \
+	.source = mco_source(DT_INST_REG_ADDR(inst)),                   \
+	.prescaler = mco_prescaler(DT_INST_REG_ADDR(inst),              \
+				   DT_INST_PROP(inst, prescaler)),      \
```
Nit 1: could you integrate the `DT_INST_` crust in the `mco_xxx` macros? I will add a suggestion on them.
```suggestion
	.source = mco_source(inst),                   \
	.prescaler = mco_prescaler(inst),             \
```

mathieuchopstm commented on drivers/clock_control/clock_stm32_mco.c:
```diff
@@ -0,0 +1,136 @@
+}
+
+#define STM32_MCO_INIT(inst)                                            \
+									\
```
It appears we lose the prescaler restriction provided by Kconfig `range`. Could you re-integrate it here?

While we _could_ add manual `#if SERIES_xxx { BUILD_ASSERT(1 <= prescaler <= XX }` checks, I think we might be able to make a generic check like:
```c
//place this before the STM32_MCO_INIT macro definition

/* Use a macro to have everything expanded in the error message
 * IS_ENABLED() here serves as a moral equivalent to C preprocessor
 * `defined()`, which is only available in #if checks
 */
#define ASSERT_VALID_PRESCALER(inst) \
        BUILD_ASSERT(IS_ENABLED(mco_prescaler(inst)),
                     "Invalid prescaler selected for MCO instance " #inst
                     " (" DT_NODE_FULL_NAME(DT_DRV_INST(inst))  ")");
```

```suggestion
									\
ASSERT_VALID_PRESCALER(inst);                                           \
```

joerchan replied:
Just to be clear, giving an invalid prescaler value would end in compilation error, because the definition does not exists, which is the same reason why your suggested IS_ENABLED also work.

This suggestion does give an explanation, so it is better.
The resulting macro expansion explanation is still horrible though.

joerchan replied:
I cannot get this suggestion to work. I will add another check, that is not as elegant, but does give a proper explanation to the user.

mathieuchopstm replied:
Good catch. Kconfig is limited to merely `range` - since it's better than nothing, I guess it was added, but did not map perfectly to all series' hardware.

My suggestion was indeed to provide a clearer error message, since, as you said, invalid prescalers will break the build regardless as the LL macro will be missing.

Since my suggestion doesn't work, ignore this comment as this is mostly cosmetic.

joerchan replied:
With the new approach, the user needs to give the divisor value for the register directly.
For each platform we can make defines for the supported values.
There is a lot of variance in what is supported or not.

joerchan replied:
Users need to go and look up the values in the reference manual though, as you can see how this looks in the sample.
The choices and values are different for every device. But can be added later, I will not do it now.

mathieuchopstm commented on drivers/clock_control/clock_stm32_mco.c:
```diff
@@ -0,0 +1,136 @@
+#endif
+#endif /* DT_NODE_HAS_STATUS(DT_NODELABEL(mco2), okay) */
+
+#define fn_mco_prescaler(i, v) LL_RCC_MCO ## i ## _DIV_ ## v
+#define mco_prescaler(i, v) fn_mco_prescaler(i, v)
+
+#define fn_mco_source(i) MCO ## i ## _SOURCE
+#define mco_source(i) fn_mco_source(i)
```
See previous comment:

```suggestion
#define fn_mco_prescaler(i, v) LL_RCC_MCO ## i ## _DIV_ ## v
#define mco_prescaler(inst) fn_mco_prescaler(DT_INST_REG_ADDR(inst),        \
                                             DT_INST_PROP(inst, prescaler))

#define fn_mco_source(i) MCO ## i ## _SOURCE
#define mco_source(inst) fn_mco_source(DT_INST_REG_ADDR(inst))
```

mathieuchopstm replied:
Possible that another layer is needed for proper expansion; I think it's fine to do so.

mathieuchopstm reviewed (changes requested):
Off with a very good start!
Some points to address or discuss:

joerchan commented on drivers/clock_control/clock_stm32_mco.c:
```diff
@@ -0,0 +1,136 @@
+const static struct stm32_mco_config stm32_mco_config_##inst = {        \
+	.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(inst),                   \
+};                                                                      \
+static struct stm32_mco_data stm32_mco_data_##inst = {                  \
+	.source = mco_source(DT_INST_REG_ADDR(inst)),                   \
+	.prescaler = mco_prescaler(DT_INST_REG_ADDR(inst),              \
+				   DT_INST_PROP(inst, prescaler)),      \
+};                                                                      \
```
I originally tried to put these in the config struct. However due to how the definitions for mco source is defined this was not possible.
The resulting comilation error is hard to understand due to the define stack-trace, but the essense of it is that source is not "const".

mathieuchopstm replied:
Intriguing. I will investigate.

mathieuchopstm replied:
Now builds fine on my part, for 3 different boards (`nucleo_f401re`, `stm32l4r9i_disco` and `nucleo_u5a5zj_q`), maybe due to the changes in the macro.

Worth trying again :smiley:

erwango commented on dts/bindings/clock/st,stm32-clock-mco.yaml:
```diff
@@ -0,0 +1,43 @@
+        e.g. clock source and prescaler.
+        Example:
+                &mco {
+                        source = <STM32_SRC_LSE>;
+                        prescaler = <2>;
```
Binding-wise, I have a preference to reuse the same binding as the usual stm32 `clocks` binding, as for instance: 
```
clocks = <&rcc STM32_SRC_PLL_Q MCO_SEL(4)>,
```
See [my comment here](https://github.com/zephyrproject-rtos/zephyr/pull/68846#discussion_r1502800920).

As an additional benefit, it would remove the whole `#if DT_PROP(DT_NODELABEL(mco1), source) == NO_SEL` section at the beginning of this file and avoids to face LL API discrepancy issues (like the `#if (CONFIG_SOC_SERIES_STM32G0X || ...` above).

The only trick would that we wouldn't be allowed to use clock_control API as MCO clocks are supposed to be configured before PLLs (as mentioned in a previous comment).

This being said, I won't block for this.

mathieuchopstm replied:
By convention, the `xxx_SEL()` macros accept the value that goes in mux selection register (corresponding here to the `LL_RCC_MCOxSOURCE_XXX` definitions).

In peripherals, it also encodes RCC offset, mask and position, but it wouldn't (shouldn't?) be necessary in this case.

joerchan replied:
I'm not sure what you mean by register offset.
But at the moment the STM32_CLOCK macro only allows 3-bit values, and the MCO selection is 4-bits.

If MCO_SEL(4) is the input to the conf register then chooses the clock, so why do we need STM32_SRC_PLL_Q  in addition?
We have 0100: HSE clock selected, then having STM32_SRC_HSE in addition means there is two values that must match.

What is recommendation here then?
Still define MCO_SEL and use raw numbers in this?

joerchan replied:
@erwango 
```
#define MCO_SEL_HSE 4
#define MCO_PRE_DIV_4 2
#define MCO_SEL(val) ((val & 0xf) << 24)
#define MCO_PRE(val) ((val & 0x3) << 28)

source = <MCO_SEL(MCO_SEL_HSE)>
prescaler = <MCO_PRE(MCO_PRE_DIV_4)>
```
Seems to be the most clear to me, the user can then easily see that the have chosen HSE divided by 4.

erwango replied:
That would work for me, thanks.

joerchan replied:
Updated approach after discord discussion.

Will use something like:
```
clocks = <&rcc STM32_SRC_PLL_Q MCO_SEL(4)>,
prescaler = <MCO_PRE(2)>;
```
Where MCO_SEL and MCO_PRE encodes values that go directly into the RCC_CFGR register.
Using STM_CLOCK_CFGR macro because of bit-width requirement for these values.

erwango commented on drivers/clock_control/clock_stm32_mco.c:
```diff
@@ -0,0 +1,136 @@
+DEVICE_DT_INST_DEFINE(inst, stm32_mco_init, NULL,                       \
+	&stm32_mco_data_##inst,                                         \
+	&stm32_mco_config_##inst,                                       \
+	PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,               \
```
This follows existing behavior so this is ok, but ideally this should happen before clock control driver initialization.
See https://github.com/zephyrproject-rtos/zephyr/issues/31912#issuecomment-1334061384

joerchan replied:
Do you mind explaining why it needs to be done in this order. The init priority is chosen so that the pinctrl apply function will work.
Because doing that before GPIO is initialized will not work.

So then we need two different priority levels.

erwango replied:
> The init priority is chosen so that the pinctrl apply function will work.

Good point.

> Do you mind explaining why it needs to be done in this order ?

According to RM, this is all about avoiding glitches.
`Clock source selection may generate glitches on MCO2.`
`Modification of this prescaler may generate glitches on MCO2`

So it depends if we say that avoiding glitches worthes the cost of having two steps of initialization for this driver. As I said, the proposed implementation is close to existing behavior in that matter, so it can remain as is. 
If required, fixing this behavior could be done later (as transparent for user).

joerchan replied:
Alright. Agree to fix it later as it is replaces the current behavior. Good to keep the changes in this PR down.

erwango reviewed (changes requested):
Thanks for the work on this long awaited piece of code.

joerchan commented on drivers/clock_control/Kconfig.stm32:
```diff
@@ -47,117 +47,17 @@ config CLOCK_STM32_MUX
 	help
 	  Use EXT_HSE as source of MCO1
 
-config CLOCK_STM32_MCO1_SRC_LSE
```
No objection to a deprecation phase, I'll update to deprecated, but will leave the final removal up to stm maintainers.

joerchan commented on include/zephyr/dt-bindings/clock/stm32f1_clock.h:
```diff
@@ -23,9 +23,15 @@
+#define STM32_SRC_HSI               (STM32_SRC_LSI + 1)
+#define STM32_SRC_HSE               (STM32_SRC_HSI + 1)
+#define STM32_SRC_PLLCLK_DIV2       (STM32_SRC_HSE + 1)
+#ifdef SOC_STM32F10X_CONNECTIVITY_LINE_DEVICE
```
Done, but since there is no stm32f10x.dts specific file then the include of the clock header had to be conditional there as well.

erwango replied:
What about stm32f105.dtsi ?

joerchan committed:
include: dt-bindings: clock: fix stm32f7 missing HSE and PLLI2S_R source

Fix stm32f7 missing the definition for HSE and PLLI2S_R clock source.

Signed-off-by: Joakim Andersson <joerchan@gmail.com>

joerchan committed:
drivers: clock_control: Expose enabled_clock for clock driver library

Expose the helper function enabled_clock so that it can be used in
other clock library sources.

Signed-off-by: Joakim Andersson <joerchan@gmail.com>

joerchan committed:
samples: stm32: mco: Add disco board which has 2 MCO instances

Add stm32f746g_disco board to MCO sample.
This board has 2 MCO instances so useful to show multiple instances
as well as checking the implementation.

Signed-off-by: Joakim Andersson <joerchan@gmail.com>

joerchan committed:
include: dt-bindings: Add MCO select and prescaler macros for CFGR reg

Add MCO select and prescaler macros to set bits in the RCC CFGRx
register.
The set of supported boards are chosen to replace what is currently
supported in Kconfig.

Signed-off-by: Joakim Andersson <joerchan@gmail.com>

joerchan committed:
boards: Add MCO support for the stm32c0xx family

Add MCO support for the stm32c0xx family.

Signed-off-by: Joakim Andersson <joerchan@gmail.com>

joerchan commented on drivers/clock_control/Kconfig.stm32:
```diff
@@ -47,117 +47,17 @@ config CLOCK_STM32_MUX
-	help
-	  MCO1 output disabled, no clock on MCO1
+config CLOCK_STM32_MCO
+	bool "STM32 clock mco driver"
```
If it has no prompt it is just an alias of DT_HAS_ST_STM32_CLOCK_MCO_ENABLED.
So i will remove this option and use DT_HAS_ST_STM32_CLOCK_MCO_ENABLED directly instead.

joerchan replied:
@erwango: Before I remove it, are you sure that you want to have it promptless?
It added it to match the behavior of the MUX:
```
config CLOCK_STM32_MUX
	bool "STM32 clock mux driver"
	default y
	depends on DT_HAS_ST_STM32_CLOCK_MUX_ENABLED
	help
```

joerchan commented on dts/arm/st/c0/stm32c0.dtsi:
```diff
@@ -67,6 +67,23 @@
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		mco1: mco@1 {
```
@erwango After the last code changes the @1 and reg =<1> is no longer needed. (only needed to macro concat before).

So question is if we should keep it or not?

mathieuchopstm commented on dts/bindings/clock/st,stm32-clock-mco.yaml:
```diff
@@ -0,0 +1,40 @@
+        Example:
+                &mco1 {
+                        clocks = <&rcc STM32_SRC_LSE MCO1_SEL(7)>;
+                        prescaler = <MCO1_DIV(1)>;
```
To be consistent with the macros provided in headers:
```suggestion
                        prescaler = <MCO1_PRE(1)>;
```

mathieuchopstm commented on include/zephyr/drivers/clock_control/stm32_clock_control.h:
```diff
@@ -16,6 +16,8 @@
 #include <zephyr/dt-bindings/clock/stm32c0_clock.h>
 #elif defined(CONFIG_SOC_SERIES_STM32F0X)
 #include <zephyr/dt-bindings/clock/stm32f0_clock.h>
+#elif defined(SOC_STM32F10X_CONNECTIVITY_LINE_DEVICE)
```
Nit: it could be wise to document that this check *must* be done before the `if defined(CONFIG_SOC_SERIES_STM32F1X)` check.

mathieuchopstm commented on samples/boards/stm32/mco/boards/nucleo_u5a5zj_q.overlay:
```diff
@@ -3,10 +3,20 @@
+ */
+#define MCO1_SEL_LSE 7
+
+/* See referance manual:
```
Typo:
```suggestion
/* See reference manual:
```

mathieuchopstm commented on samples/boards/stm32/mco/boards/stm32f746g_disco.overlay:
```diff
@@ -0,0 +1,44 @@
+ */
+#define MCO1_SEL_LSE 1
+
+/* See referance manual:
```
Typo:
```suggestion
/* See reference manual:
```

mathieuchopstm commented on samples/boards/stm32/mco/boards/stm32f746g_disco.overlay:
```diff
@@ -0,0 +1,44 @@
+ */
+#define MCO2_SEL_HSE 2
+
+/* See referance manual:
```
Typo:
```suggestion
/* See reference manual:
```

Nit: `RMxxxx` with appropriate RM number would be more appreciated (but it's not a big deal)

mathieuchopstm reviewed (changes requested):
Looks pretty solid!

A few minor points to iron out:

joerchan committed:
doc: Document deprecated stm32 MCO kconfig options for MCO

Document deprecated stm32 MCO Kconfig options in favour of
setting MCO through devicetree.

Signed-off-by: Joakim Andersson <joerchan@gmail.com>

erwango reviewed (changes requested):
Few more changes required but a great thanks for this new version.

mathieuchopstm commented on include/zephyr/dt-bindings/clock/stm32f10x_clock.h:
```diff
@@ -72,4 +73,8 @@
 #define RTC_SEL(val)		STM32_CLOCK(val, 3, 8, BDCR_REG)
 
+#define MCO1_SEL(val)           STM32_MCO_CFGR(val, 0x7, 24, CFGR1_REG)
+/* Prescaler not supported. */
```
I think the intent is to make explicit that "`MCO1_PRE` is not present because the feature is not available, not because it was forgotten".

joerchan replied:
This series does not have a MCOPRE in CFGR register, the mask is set to zero.

joerchan replied:
Should we improve on the comment? Happy to take suggestions here.

mathieuchopstm replied:
Proposal:
```suggestion
/* STM32F1 Connectivity line has no MCO prescaler. */
```

Although, I see now that the `prescaler` property is marked as `required: true`, so we have to handle this special case.

I see two solutions:
* remove the `required` attribute from the property, and use `DT_PROP_OR(node, prescaler, /* DRIVER_SENTINEL_DEFINE */)` when obtaining prescaler
  * when applying the configuration, only write the prescaler if `config->prescaler != /* DRIVER_SENTINEL_DEFINE */`
  * define could be something like `#define MCO_DONT_CONFIGURE_PRESCALER 0xFFFFFFFF` in `clock_stm32_mco.c`
* keep the `required` attribute on the property, but define a (universal) fallback value (e.g., in `stm32_clocks_common.h`, `#define STM32_MCO_CFGR_NONE 0xFFFFFFFF`)
  * in `stm32f10x_clock.h`, we'd have a single `#define MCO1_NO_PRESCALER STM32_MCO_CFGR_NONE` define in place of `MCO_PRE`
  * driver would still only apply configuration if `config->prescaler != STM32_MCO_CFGR_NONE`

I would lean towards the second solution, which could give something easy to understand like:
```suggestion
/* STM32F1 Connectivity line has no MCO prescaler.
 * Use the following as the `prescaler` property for MCO1:
 */
#define MCO1_PRESCALER_NONE STM32_MCO_CFGR_NONE
```

It's also safer as it guarantees the user has explicitly chosen the prescaler - no possible bad surprise.

CC'ing @erwango for input

erwango replied:
To me, the canonical way (maybe not the most simple) to deal with this is:
- dts/bingings: create a new `st,stm32f1-mco` compatible which includes `st,stm32-mco` but blocks `prescaler` property.
- dts/arm/st... use the correct compatible
- driver/clock_control
 - Kconfig: depends on DT_HAS_ST_STM32_MCO_ENABLED || DT_HAS_ST_STM32F1_MCO_ENABLED
 - ...._mco.c: 
   - Define `DT_DRV_COMPAT` conditionally
```
 #if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_mco)
#define DT_DRV_COMPAT st_stm32_mco
#else
#define DT_DRV_COMPAT st_stm32f1_mco
#endif
```    
   - Encapsulate the prescaler setting under `#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_mco)`
   - Define prescaler with 
```
	IF_ENABLED(DT_HAS_COMPAT_STATUS_OKAY(st_stm32_mco),		\
		(.prescaler = DT_INST_PROP(id, prescaler),))		\
```

joerchan replied:
Personally I think leaving the prescaler as not required would not be a bad idea.

If it is not used, then we can just leave it at its register default value, this can be added to the property description.
I think across all devices the reset value of the CFGR register results in a prescaler division of 1 i.e. no division.
If I understand correctly this is Mathiu's first option.

We can update the sample to also demonstrate this, for example on the discovery board which has 2 MCOs, we can have one with prescaler, and one without.

In the driver we throw an error for the devices that don't have prepscaler support if this is added.
This can be done by checking for Kconfig F1xx series, or we can have devicetree property like: `no-prescaler-support` added on this series.

@erwango  I guess you're the one to make a decision?
Even with your approach we may also consider leaving prescaler as optional and just leave it at reset value.
Two things to be decided:
1. How or if to block prescaler configuration on f1xx series.
2. Should prescaler be required.

mathieuchopstm replied:
> Personally I think leaving the prescaler as not required would not be a bad idea.

This is what Erwan is proposing, but using a specific compatible for the lines/series with such quirk (i.e., STM32F1).

This way, F1 DTS will not require `prescaler` (and, in fact, error out if provided), but all other series will. This ensures that on all series, the MCO is fully configured by DTS (and thus user) before being enabled.

> If it is not used, then we can just leave it at its register default value, this can be added to the property description.
I think across all devices the reset value of the CFGR register results in a prescaler division of 1 i.e. no division.
If I understand correctly this is Mathiu's first option.

Actually, there are exceptions to this - for example, on STM32H7 (RM0399): `MCO2 prescaler [...] 0000: prescaler disabled (default after reset)` - i.e., default prescaler configuration is *disable MCO*, not *divide by 1*.

> How or if to block prescaler configuration on f1xx series.

If you mean, how to declare it in a binding, see e.g.: https://github.com/zephyrproject-rtos/zephyr/blob/494f88070f6bb909389630264e4f89c71072ae53/dts/bindings/clock/st%2Cstm32f100-pll-clock.yaml#L25-L28

In our case, it would be `st,stm32f1-mco.yaml` containing:
```yaml
 include: 
   - name: st,stm32-mco.yaml 
     property-blocklist: 
       - prescaler
```

cc @erwango

erwango replied:
> This is what Erwan is proposing

Exactly:
- we let dt thow errors on misconfigurations (eg prescaler on non compatible F1 devices)
- in the driver we only deal with correct configurations

joerchan replied:
Alright, going for that approach then.

joerchan commented on drivers/clock_control/clock_stm32_mco.c:
```diff
@@ -0,0 +1,82 @@
+	const struct pinctrl_dev_config *pcfg;
+	uint32_t prescaler;
+	/* clock subsystem driving this peripheral */
+	const struct stm32_pclken *pclken;
```
I didn't see this method used anywhere else. Isn't it better to keep it consistent here, at the cost of one pointer?

mathieuchopstm replied:
Fine by me either way.

joerchan replied:
Alright. I put your suggestion in.

joerchan commented on samples/boards/stm32/mco/boards/nucleo_u5a5zj_q.overlay:
```diff
@@ -3,10 +3,20 @@
-		pinctrl-0 = <&rcc_mco_pa8>;
-		pinctrl-names = "default";
-	};
+/* See reference manual:
+ *   0111: LSE clock selected
+ */
+#define MCO1_SEL_LSE 7
```
I agree, but will skip for now.

mathieuchopstm commented on doc/releases/release-notes-4.0.rst:
```diff
@@ -101,6 +101,8 @@ Boards & SoC Support
   * :ref:`native_posix<native_posix>` has been deprecated in favour of
     :ref:`native_sim<native_sim>`.
   * Support for Google Kukui EC board (``google_kukui``) has been dropped.
+  * STM32: Deprecated setting MCO source and prescalor in Kconfig in favour of setting it through devicetree.
+    See ``samples/boards/stm32/mco`` sample.
```
Nit:
```suggestion
  * STM32: Deprecated MCO configuration via Kconfig in favour of setting it through devicetree.
    See ``samples/boards/stm32/mco`` sample.
```

mathieuchopstm commented on samples/boards/stm32/mco/boards/stm32f746g_disco.overlay:
```diff
@@ -0,0 +1,44 @@
+ */
+#define MCO2_SEL_HSE 2
+
+/* See reference manual (RM0385):
+ *   111: division by 5
+ */
+#define MCO2_PRE_DIV_NONE 7
```
Why is the macro named `DIV_NONE` if this selects `prescaler = 5`?

joerchan replied:
Because I changed my mind and didn't update it properly. (debugging something that didn't work and wanted to see if the frequency was too high for the pin or my logic analyzer).

mathieuchopstm reviewed (commented):
Minor things, otherwise LGTM besides the remaining open discussion points.

mathieuchopstm commented on include/zephyr/drivers/clock_control/stm32_clock_control.h:
```diff
@@ -17,7 +17,11 @@
 #elif defined(CONFIG_SOC_SERIES_STM32F0X)
 #include <zephyr/dt-bindings/clock/stm32f0_clock.h>
 #elif defined(CONFIG_SOC_SERIES_STM32F1X)
+#ifdef defined(CONFIG_SOC_STM32F10X_CONNECTIVITY_LINE_DEVICE)
```
Good catch by CI:
```suggestion
#if defined(CONFIG_SOC_STM32F10X_CONNECTIVITY_LINE_DEVICE)
```

mathieuchopstm reviewed (changes requested):
A few other things found after reading everything a second time:

joerchan commented on include/zephyr/dt-bindings/clock/stm32f10x_clock.h:
```diff
@@ -0,0 +1,82 @@
+/* Low speed clocks defined in stm32_common_clocks.h */
+#define STM32_SRC_HSI               (STM32_SRC_LSI + 1)
+#define STM32_SRC_HSE               (STM32_SRC_HSI + 1)
+#define STM32_SRC_PLLCLK_DIV2       (STM32_SRC_HSE + 1)
+#define STM32_SRC_EXT_HSE           (STM32_SRC_PLLCLK_DIV2 + 1)
+#define STM32_SRC_PLL2CLK           (STM32_SRC_EXT_HSE + 1)
+#define STM32_SRC_PLLI2SCLK         (STM32_SRC_PLL2CLK + 1)
+#define STM32_SRC_PLLI2SCLK_DIV2    (STM32_SRC_PLLI2SCLK + 1)
```
I added these based on Kconfig, but I cannot seem to find a reference manual that lists these for MCO select field.
Are these duplications of other definitions used?

These pairs look similar.

PLLI2SCLK
PLLI2S

PLL2PCLK
PLL2CLK

HSE
EXT_HSE

The _DIV2 I understand are special for the f1x / f10x lines as they don't have the PRESCALER
It may look like the MCO is always divided by two, and so maybe we don't need the _DIV2 to check that the clock source is enabled?

These definitions were added to make the macro expansion solution work properly, but we're not using that approach anymore.

mathieuchopstm replied:
Going one by one:
* `PLLI2SCLK` is a misnomer for `PLL3` that (presumably) originated from STM32CubeF1.

https://github.com/zephyrproject-rtos/hal_stm32/blob/808602c497998f9e84f042ef90cf2010d69789dd/stm32cube/stm32f1xx/drivers/include/stm32f1xx_ll_rcc.h#L250-L272
```c
#define LL_RCC_MCO1SOURCE_PLLI2SCLK_DIV2   RCC_CFGR_MCO_PLL3CLK_DIV2 /*!< PLLI2S clock divided by 2 selected as MCO source*/
#define LL_RCC_MCO1SOURCE_PLLI2SCLK        RCC_CFGR_MCO_PLL3CLK      /*!< PLLI2S clock selected as MCO source */
```

Since we're dropping LL macros, I think it would be wiser to match RM and use `STM32_SRC_PLL3` - as it doesn't exist yet, support for it must be added.

* `PLLI2S` is a real PLL block on STM32F4/STM32F7

RM0368 makes it clear that MCO corresponds to PLLI2S R output, so the existing `STM32_SRC_PLLI2S_R` can be used for F4.

An identically named  macro should be introduced for F7 so it can be used there as well (RM0431 is *less explicit* about it, but `PLLI2S` MCO source should be `R` output on F7 too). 

* `PLL2PCLK` is `PLL2` `P` output on STM32H5/H7, so it can be mapped to existing `STM32_SRC_PLL2_P`

* `PLL2CLK` is the (single) `PLL2` output on STM32F1, so it should be mapped to a separate `STM32_SRC_PLL2` macro, for which support would need to be added.

* For the `_DIV2` sources, since we are not allowing `clock_control_get_subsys_rate` queries on the MCO sources for now, the regular and `_DIV2` sources can use the same `STM32_SRC_xxx` macro.
  * For `STM32_SRC_PLLCLK_DIV2` specifically, since no non-`_DIV2` equivalent exists, I think the macro can remain as `STM32_SRC_PLLCLK_DIV2` though.

joerchan replied:
If I build for one of the connectivity lines i get the following options for MCO1 output:
`west build -b stm3210c_eval samples/boards/stm32/mco -t menuconfig`
```
(X) NOCLOCK
( ) EXT_HSE
( ) HSE
( ) HSI
( ) PLLCLK_DIV2
( ) PLL2CLK
( ) PLLI2SCLK
( ) PLLI2SCLK_DIV2
( ) SYSCLK
```

However looking up STM32F107VC SoC gives RM008 as the reference manual, which only has these options as MCOSEL:
```
SYSCLK
HSI
HSE
PLL clock divided by 2
```

So either I don't have the right reference manual, or the Kconfig selection for this board is wrong.

mathieuchopstm replied:
The STM32F107VC is a device of the Connectivity Line (`Connectivity line devices are STM32F105xx and STM32F107xx microcontrollers.`). Are you sure that you looked at Chapter 8 "`Connectivity line devices: RCC`" of RM0008, and not Chapter 7 "`Low-, medium-, high- and XL-density RCC`"?

Quoting page 135 of RM0008 Rev 21 (corresponding Kconfig in `[]`):
```
Bits 26:24 MCO[3:0]: Microcontroller clock output
Set and cleared by software.
00xx: No clock                                                        [NOCLOCK]
0100: System clock (SYSCLK) selected                                  [SYSCLK]
0101: HSI clock selected                                              [HSI]
0110: HSE clock selected                                              [HSE]
0111: PLL clock divided by 2 selected                                 [PLLCLK_DIV2]
1000: PLL2 clock selected                                             [PLL2CLK]
1001: PLL3 clock divided by 2 selected                                [PLLI2SCLK_DIV2]
1010: XT1 external 3-25 MHz oscillator clock selected (for Ethernet)  [EXT_HSE]
1011: PLL3 clock selected (for Ethernet)                              [PLLI2SCLK]
```

joerchan replied:
Thanks. I didn't know it was two different chapters for the RCC configuration. Then it makes sense.

joerchan replied:
So, from what I can tell, what is missing is:
F10x (connectivity line) support for PLL3 clock source.
F7 support for PLLI2S_R clock source.

I will add:
PLL2 and PLL3 support to `clock_stm32_ll_common.c:enabled_clock()` function, and PLL2/3 clock source definition to f10x, and PLLI2S_R to f7.

At that point the DTS approach will be feature-complete compared to what the Kconfig approach had, so deprecation can happen, and the missing support can be added as a follow-up to this PR.

erwango replied:
> At that point the DTS approach will be feature-complete compared to what the Kconfig approach had, so deprecation can happen, and the missing support can be added as a follow-up to this PR (by someone else)

Excellent. Thanks again @joerchan

joerchan replied:
Adding STM32_SRC_PLLCLK to stm32f1 devices cannot be done without also implementing `get_pllout_frequency` in `clock_stm32f1.c`.
However this seems to be the right way to refer to this node.

Is there an easy way to tell the frequency for this PLL? (looking at clock_f0_f3.c it looks a bit complicated, and like it needs to be a variant of that function).

joerchan committed:
drivers: clock_control: Add clock sources to common enabled_clock check

Add clock sources PLL2CLK, PLL3CLK and EXT_HSE.
Needed to check that these clocks are enabled in MCO code.

Signed-off-by: Joakim Andersson <joerchan@gmail.com>

mathieuchopstm commented on drivers/clock_control/clock_stm32f1.c:
```diff
@@ -23,6 +23,11 @@
 #if defined(STM32_PLL_ENABLED)
 
+uint32_t get_pllout_frequency(void)
+{
```
A comment would be appreciated.
```suggestion
{
    /* Stub implementation for compatibility with clock_stm32_ll_common.
     * The PLL domain clock is only used for MCO configuration, but the
     * MCO driver never queries the PLL output clock frequency.
     */
```

mathieuchopstm commented on dts/bindings/clock/st,stm32-clock-mco.yaml:
I think the bindings should be added to the tree in a separate commit.

mathieuchopstm commented on samples/boards/stm32/mco/src/main.c:
```diff
@@ -7,15 +7,8 @@
-	/* Configure the MCO pin using pinctrl in order to set the alternate function of the pin. */
-	const struct pinctrl_dev_config *pcfg = PINCTRL_DT_DEV_CONFIG_GET(DT_PATH(zephyr_user));
-	(void)pinctrl_apply_state(pcfg, PINCTRL_STATE_DEFAULT);
-
```
Nit: sample could check that the MCO devices have been successfully initialized.

At least, a comment explaining why nothing is done in the sample source code would be appreciated.

mathieuchopstm reviewed (commented):
Minor issues, a few nits, but overall looks complete.

erwango commented on include/zephyr/dt-bindings/clock/stm32f10x_clock.h:
```diff
@@ -0,0 +1,75 @@
+/*
```
Non blocking: Cf `include/zephyr/dt-bindings/clock/stm32f427_clock.h` to see how we can minimize code duplication in these files in similar situation

erwango replied:
> I don't understand what "Cf" means.

Sorry, "Cf" stands for `confere` as "Please see". 

> Looks odd, but maybe intentional?

That's indeed as expected (reusing existing binding when compatibles).
But I agree this is not exactly what I though as in this case as base  `stmf324_clock.h` (which contain all common pieces) is used and then complementary ` stm32f410_clock.h` or ` stm32f427_clock.h` are added on matching devices (which then also benefits to the SoCs that inlcudes those .dtsi, as stm32f429.dtsi for instance) to add the variant specific bindings.

joerchan commented on dts/arm/st/f1/stm32f1.dtsi:
```diff
@@ -6,7 +6,11 @@
  */
 
 #include <arm/armv7-m.dtsi>
+#if defined(SOC_STM32F10X_CONNECTIVITY_LINE_DEVICE)
+#include <zephyr/dt-bindings/clock/stm32f10x_clock.h>
+#else
 #include <zephyr/dt-bindings/clock/stm32f1_clock.h>
+#endif
```
Thanks. Doing this means that both get included for connectivity line and MCO1_SEL needs to be redefined, I hope that is an acceptable thing to do.

joerchan committed:
dts: bindings: Add stm32 microcontroller clock output binding

Add stm32 microcontroller clock output binding.

Signed-off-by: Joakim Andersson <joerchan@gmail.com>

joerchan committed:
include: Add STM32_MCO_CFGR bit encoding and decoding

Add macro for encoding MCO bit values into the CFGR register
of RCC. Add macros for decoding and extracting the different fields.

Signed-off-by: Joakim Andersson <joerchan@gmail.com>

joerchan commented on drivers/clock_control/Kconfig.stm32:
```diff
@@ -46,7 +46,13 @@ config CLOCK_STM32_MUX
+	bool
+	default y
+	depends on DT_HAS_ST_STM32_CLOCK_MCO_ENABLED || DT_HAS_ST_STM32F1_CLOCK_MCO_ENABLED
+	# Device tree configuration is ignored when MCO source symbols are set.
```
reworded the second line a bit.

joerchan commented on dts/bindings/clock/st,stm32f1-clock-mco.yaml:
```diff
@@ -0,0 +1,34 @@
+compatible: "st,stm32f1-clock-mco"
+
+description: |
+        STM32 F1 series Microcontroller clock output (MCO)
+        Configure the microcontroller clock output pin and additional properties,
+        e.g. clock source.
```
I took the explanation, but split it up, so that first part describes the first example, and second part describes the second example.

mathieuchopstm reviewed (changes requested):
Minor details in a few spots.
`-1` for the improper `IS_ENABLED` usage, but otherwise LGTM.

joerchan commented:
I will rebase this PR and resolve the conflict after the review is finished, so that it is easier to see the updates that are being done.

The conflict is trivial to resolve, only adjacent context lines that have been updated.

mathieuchopstm reviewed (dismissed):
One nit from last review remaining, but non-blocking. LGTM otherwise.

joerchan commented:
> One nit from last review remaining, but non-blocking. LGTM otherwise.

Addressed last nit and rebased the PR to fix the conflicts.

joerchan commented:
Re-running documentation build, failed because of server issues.

erwango reviewed (approved):
Once again a great thanks!
